main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -371,6 +373,195 @@ object Inliner {
+
+  private object MatchResult {
+
+    case class Match(binders: Chain[(Option[Pattern.Var], MonoAst.Expr)]) extends MatchResult
+
+    case object NoMatch extends MatchResult
+
+    case object Unknown extends MatchResult
+
+    def emptyMatch(): MatchResult =
+      Match(Chain.empty)
+
+    def singleMatch(pat: Option[Pattern.Var], exp: MonoAst.Expr): MatchResult =
+      Match(Chain((pat, exp)))
+
+    def matchFromBool(b: Boolean): MatchResult =
+      if (b) emptyMatch() else NoMatch
+
+    /**
+      * Concatenates two match results.
+      *
+      * If either result is [[NoMatch]], the output is [[NoMatch]].
+      *
+      * Then if either result is [[Unknown]], the output is [[Unknown]].
+      *
+      * Then (both results now being [[Match]]) the binders are concatenated.
+      */
+    def concat(mr1: MatchResult, mr2: MatchResult): MatchResult = (mr1, mr2) match {
+      case (MatchResult.NoMatch, _) | (_, MatchResult.NoMatch) =>
+        MatchResult.NoMatch
+      case (MatchResult.Unknown, _) | (_, MatchResult.Unknown) =>
+        MatchResult.Unknown
+      case (MatchResult.Match(binders1), MatchResult.Match(binders2)) =>
+        MatchResult.Match(binders1 ++ binders2)
+    }
+
+  }
+
+  /** Returns the match result of `exp` against `pat` with the guard `guardOpt`. */
+  private def matchRule(exp: MonoAst.Expr, pat: Pattern, guardOpt: Option[MonoAst.Expr]): MatchResult = {
+    val guardVal = guardOpt.map(evalBoolExpression).getOrElse(FuzzyBool.True)
+    (matchPat(exp, pat), guardVal) match {
+      case (_, FuzzyBool.False) | (MatchResult.NoMatch, _) =>
+        MatchResult.NoMatch
+      case (MatchResult.Unknown, _) | (MatchResult.Match(_), FuzzyBool.Unknown) =>
+        MatchResult.Unknown
+      case (resultMatch@MatchResult.Match(_), FuzzyBool.True) =>
+        resultMatch
+    }
+  }
+
+  /** Returns the match result of `exp` against `pat`. */
+  private def matchPat(exp: MonoAst.Expr, pat: MonoAst.Pattern): MatchResult = pat match {
+    case Pattern.Wild(_, _) =>
+      // Preserve the expression in case it is impure.
+      MatchResult.singleMatch(None, exp)
+
+    case v@Pattern.Var(_, _, _, _) =>
+      MatchResult.singleMatch(Some(v), exp)
+
+    case Pattern.Cst(cst, _, _) =>
+      exp match {
+        case Expr.Cst(expCst, _, _) =>
+          matchConstant(cst, expCst)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Tag(symUse, pats, _, _) =>
+      exp match {
+        case Expr.ApplyAtomic(AtomicOp.Tag(caseSym), exps, _, _, _) =>
+          if (symUse.sym != caseSym) MatchResult.NoMatch
+          else matchPatterns(exps, pats)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Tuple(pats, _, _) =>
+      exp match {
+        case Expr.ApplyAtomic(AtomicOp.Tuple, exps, _, _, _) =>
+          matchPatterns(exps, pats.toList)
+        case _ =>
+          MatchResult.Unknown
+      }
+
+    case Pattern.Record(_, _, _, _) =>
+      MatchResult.Unknown
```

Comment:
Too bad; we have a lot of labelled arguments where this might help :)

Reply:
wanted to keep the PR small, the concrete matching precision can always be increased.